packages/react-vapor/src/components/editor/JSONEditor.tsx

```diff
@@ -5,73 +5,76 @@ import * as React from 'react';
 
 import {Svg} from '../svg/Svg';
 import {CodeEditor} from './CodeEditor';
-import {CodeMirrorModes} from './EditorConstants';
+import {CodeMirrorModes, DEFAULT_JSON_ERROR_MESSAGE} from './EditorConstants';
+import {JSONEditorUtils} from './JSONEditorUtils';
 
 export interface IJSONEditorProps {
+    id?: string;
     value: string;
     readOnly?: boolean;
     onChange?: (json: string, inError: boolean) => void;
     errorMessage?: string;
+    containerClasses?: string[];
+    className?: string;
+    ref?: React.Ref<any>;
 }
 
-export interface IJSONEditorState {
-    isInError: boolean;
+export interface IJSONEditorStateProps {
+    value: string;
 }
 
-export const DEFAULT_JSON_ERROR_MESSAGE: string = 'The JSON configuration is syntactically invalid.';
+export interface IJSONEditorDispatchProps {
+    onMount?: () => void;
+    onChange?: (json: string, inError: boolean) => void;
+    onUnmount?: () => void;
+}
 
-export class JSONEditor extends React.Component<IJSONEditorProps, IJSONEditorState> {
-    static defaultProps: Partial<IJSONEditorProps> = {
-        errorMessage: DEFAULT_JSON_ERROR_MESSAGE,
-    };
+export const JSONEditor: React.FunctionComponent<IJSONEditorProps &
+    IJSONEditorStateProps &
+    IJSONEditorDispatchProps> = ({
+    value,
+    readOnly,
+    onChange,
+    errorMessage,
+    containerClasses,
+    className,
+    onMount,
+    onUnmount,
+    ref,
+}) => {
+    const [isInError, setIsInError] = React.useState(false);
 
-    constructor(props: IJSONEditorProps, state: IJSONEditorState) {
-        super(props, state);
-        this.state = {
-            isInError: false,
-        };
-    }
+    React.useEffect(() => {
+        onMount?.();
 
-    render() {
-        const classes: string = classNames('form-group', {
-            'input-validation-error': this.state.isInError,
-        });
-        return (
-            <div className={classes}>
-                <CodeEditor
-                    value={this.props.value}
-                    onChange={(json: string) => this.handleChange(json)}
-                    mode={CodeMirrorModes.JSON}
-                    readOnly={this.props.readOnly}
-                />
-                {this.getValidationDetails()}
-            </div>
-        );
-    }
+        return onUnmount;
+    }, []);

```

Comment:
`[onMount, onUnmount]` ?

Reply:
I think that it works correctly here without dependencies, since we only need to call `onMount` and `onUnmount` once when rendering and destroying. If these props change, then it'll be another new component. Also, we've already implemented `onMount` and `onUnmount` in this way several times, e.g. https://github.com/coveo/react-vapor/blob/master/packages/react-vapor/src/components/textarea/TextArea.tsx#L78